[Github] Improve formating of PR diffs in bot notifications

cor3ntin opened the pull request:
* This avoid pinging folks on all issue when they got pinged on bugzilla eons ago
* Avoid formatting bugs when there is html in the issue description
* Truncate the list of files and the diff independently of each other. This avoids truncating cutting a file line in 2 and to cut in the middle of html markup. This is a fringe case but it does happen when people accidentally push weird branches conflicting on all the files.

Test: This is #66118 @cor3ntin

cor3ntin committed:
[Github] Escape `@` and html in the <details> block

* This avoid pinging folks on all issue when they got pinged on
bugzilla eaons ago

* Avoid formating bugs when there is html in the issue description

cor3ntin committed:
Fix order of operations: we should escape html first so the comment does not get escaped

Endilll commented on llvm/utils/git/github-automation.py:
```diff
@@ -45,6 +46,10 @@ def _get_curent_team(team_name, teams) -> Optional[github.Team.Team]:
+def escape_description(str):
+    # https://github.com/github/markup/issues/1168#issuecomment-494946168
+    str = html.escape(str, False)
+    return str.replace("@", "@<!-- -->")
```
I guess this hides names completely.
Not that it matters much, but removing `@` seems a less intrusive way.

cor3ntin replied:
Nah, it should keep everything @<!-- -->Endilll

Endilll replied:
Ah, so you're inserting a dummy tag to break @ from name.
That's smart!

cor3ntin committed:
Formatting

MaskRay commented:
I wonder whether `#` needs to be escaped as well

cor3ntin committed:
* Escape #
* Escape PR description
* Trunkate the list of files if it's > 20K
* Color the diff on github

cor3ntin committed:
Formatting

cor3ntin commented:
<pre lang="diff">
+ class {patch}
- bar
</pre>

> I wonder whether `#` needs to be escaped as well

Yes, done!

joker-eph commented on llvm/utils/git/github-automation.py:
```diff
@@ -113,6 +122,11 @@ def run(self) -> bool:
+        # and the file list to 20kB each.
+        STAT_LIMIT = 20 * 1024
+        DIFF_LIMIT = 20 * 1024
+
```
This isn't in the PR description, can you add it there? (with some context for the "why" of the change)

joker-eph reviewed (approved):
LG, but I agree with @MaskRay about `#`, it is likely we don't want to escape it alone, but the sequence of `#[1-2]+` maybe?

cor3ntin commented:
@joker-eph I'll rebase and try to be more selective in what we escape. escaping unconditionally probably does not hurt though

MaskRay commented:
Unconditionally escaping `#` mitigates problems such as a stack trace accidentally references `#1`/`#2`/etc. I think for intentional references, people can use regular comments.

joker-eph commented:
> Unconditionally escaping # mitigates problems such as a stack trace accidentally references #1/#2/etc. I think for intentional references, people can use regular comments.

You missed my point: I commented that we should match `#` **followed by at least one number. The case you mentioned would be covered just fine.

MaskRay commented:
> > Unconditionally escaping # mitigates problems such as a stack trace accidentally references #1/#2/etc. I think for intentional references, people can use regular comments.
> 
> You missed my point: I commented that we should match `#` **followed by at least one number. The case you mentioned would be covered just fine.

I'd argue that issue/PR links can be achieved with normal comments, not with bot comments. So I'd argue we should not allow `#123` to link an issue, either. This is probably not going to be useful.

cor3ntin committed:
Only escape # and @ when they could be part of an issue number/handle

cor3ntin committed:
Use a markdown codeblock for the diff.

This way we can render < and & properly.

This only work if there is a new line before the 3 ticks.

cor3ntin commented:
<details>
<summary>Spoiler</summary>

```diff
+ You need a new line before the ``` 
- `[]<typename>(auto&)`
```
</details>

cor3ntin committed:
formatting

cor3ntin committed:
formatting again using the correct base revision

tstellar commented:
Can you test this out by running the script against this PR?

cor3ntin commented:
<!--LLVM PR SUMMARY COMMENT-->

@llvm/pr-subscribers-github-workflow

<details>
<summary>Changes</summary>
* This avoid pinging folks on all issue when they got pinged on bugzilla eons ago
* Avoid formatting bugs when there is html in the issue description
* Truncate the list of files and the diff independently of each other. This avoids truncating cutting a file line in 2 and to cut in the middle of html markup. This is a fringe case but it does happen when people accidentally push weird branches conflicting on all the files.


--
Full diff: https://github.com/llvm/llvm-project/pull/66118.diff


1 Files Affected:

- (modified) llvm/utils/git/github-automation.py (+45-15) 


```diff
diff --git a/llvm/utils/git/github-automation.py b/llvm/utils/git/github-automation.py
index eac5816b5499f6a..7f4ea46a0741249 100755
--- a/llvm/utils/git/github-automation.py
+++ b/llvm/utils/git/github-automation.py
@@ -47,6 +47,21 @@ def _get_curent_team(team_name, teams) -> Optional[github.Team.Team]:
     return None
 
 
+def escape_description(str):
+    # https://github.com/github/markup/issues/1168#issuecomment-494946168
+    str = html.escape(str, False)
+    # '@' followed by alphanum is a user name
+    str = re.sub("@(?=\w+)", "@<!-- -->", str)
+    # '#' followed by digits is considered an issue number
+    str = re.sub("#(?=\d+\s)", "#<!-- -->", str)
+    return str
+
+
+def sanitize_markdown_code_block(str):
+    # remove codeblocks terminators
+    return re.sub("^\s*```\s*$", r"` ` `", str)
+
+
 class IssueSubscriber:
     @property
     def team_name(self) -> str:
@@ -67,12 +82,15 @@ def run(self) -> bool:
         if team.slug == "issue-subscribers-good-first-issue":
             comment = "{}\n".format(beginner_comment)
 
-        comment = (
-            f"@llvm/{team.slug}"
-            + "\n\n<details>\n"
-            + f"{self.issue.body}\n"
-            + "</details>"
-        )
+        body = escape_description(self.issue.body)
+
+        comment = f"""
+@llvm/{team.slug}
+
+<details>
+{body}
+</details>
+"""
 
         self.issue.create_comment(comment)
         return True
@@ -113,6 +131,11 @@ def run(self) -> bool:
             print(f"couldn't find team named {self.team_name}")
             return False
 
+        # GitHub limits comments to 65,536 characters, let's limit the diff
+        # and the file list to 20kB each.
+        STAT_LIMIT = 20 * 1024
+        DIFF_LIMIT = 20 * 1024
+
         # Get statistics for each file
         diff_stats = f"{self.pr.changed_files} Files Affected:\n\n"
         for file in self.pr.get_files():
@@ -125,35 +148,42 @@ def run(self) -> bool:
             if file.status == "renamed":
                 print(f"(from {file.previous_filename})")
             diff_stats += "\n"
-        diff_stats += "\n"
+            if len(diff_stats) > STAT_LIMIT:
+                break
 
         # Get the diff
         try:
-            patch = html.escape(requests.get(self.pr.diff_url).text)
+            patch = requests.get(self.pr.diff_url).text
         except:
             patch = ""
-        diff_stats += "\n<pre>\n" + html.escape(patch)
 
-        # GitHub limits comments to 65,536 characters, let's limit the diff to 20kB.
-        DIFF_LIMIT = 20 * 1024
+        patch = sanitize_markdown_code_block(patch)
+
         patch_link = f"Full diff: {self.pr.diff_url}\n"
         if len(patch) > DIFF_LIMIT:
             patch_link = f"\nPatch is {human_readable_size(len(patch))}, truncated to {human_readable_size(DIFF_LIMIT)} below, full version: {self.pr.diff_url}\n"
-            diff_stats = diff_stats[0:DIFF_LIMIT] + "...\n<truncated>\n"
-        diff_stats += "</pre>"
+            patch = patch[0:DIFF_LIMIT] + "...\n[truncated]\n"
         team_mention = "@llvm/{}".format(team.slug)
 
-        body = self.pr.body
+        body = escape_description(self.pr.body)
+        # Note: the comment is in markdown and the code below
+        # is sensible to line break
         comment = f"""
 {self.COMMENT_TAG}
 {team_mention}
-            
+
 <details>
 <summary>Changes</summary>
 {body}
 --
 {patch_link}
+
 {diff_stats}
+
+```diff
+{patch}
+```
+
 </details>
 """
 

```

</details>

cor3ntin commented:
<!--LLVM PR SUMMARY COMMENT-->

@llvm/pr-subscribers-github-workflow

<details>
<summary>Changes</summary>

* This avoid pinging folks on all issue when they got pinged on bugzilla eons ago
* Avoid formatting bugs when there is html in the issue description
* Truncate the list of files and the diff independently of each other. This avoids truncating cutting a file line in 2 and to cut in the middle of html markup. This is a fringe case but it does happen when people accidentally push weird branches conflicting on all the files.


--
Full diff: https://github.com/llvm/llvm-project/pull/66118.diff


1 Files Affected:

- (modified) llvm/utils/git/github-automation.py (+45-15) 


```diff
diff --git a/llvm/utils/git/github-automation.py b/llvm/utils/git/github-automation.py
index eac5816b5499f6a..7f4ea46a0741249 100755
--- a/llvm/utils/git/github-automation.py
+++ b/llvm/utils/git/github-automation.py
@@ -47,6 +47,21 @@ def _get_curent_team(team_name, teams) -> Optional[github.Team.Team]:
     return None
 
 
+def escape_description(str):
+    # https://github.com/github/markup/issues/1168#issuecomment-494946168
+    str = html.escape(str, False)
+    # '@' followed by alphanum is a user name
+    str = re.sub("@(?=\w+)", "@<!-- -->", str)
+    # '#' followed by digits is considered an issue number
+    str = re.sub("#(?=\d+\s)", "#<!-- -->", str)
+    return str
+
+
+def sanitize_markdown_code_block(str):
+    # remove codeblocks terminators
+    return re.sub("^\s*```\s*$", r"` ` `", str)
+
+
 class IssueSubscriber:
     @property
     def team_name(self) -> str:
@@ -67,12 +82,15 @@ def run(self) -> bool:
         if team.slug == "issue-subscribers-good-first-issue":
             comment = "{}\n".format(beginner_comment)
 
-        comment = (
-            f"@llvm/{team.slug}"
-            + "\n\n<details>\n"
-            + f"{self.issue.body}\n"
-            + "</details>"
-        )
+        body = escape_description(self.issue.body)
+
+        comment = f"""
+@llvm/{team.slug}
+
+<details>
+{body}
+</details>
+"""
 
         self.issue.create_comment(comment)
         return True
@@ -113,6 +131,11 @@ def run(self) -> bool:
             print(f"couldn't find team named {self.team_name}")
             return False
 
+        # GitHub limits comments to 65,536 characters, let's limit the diff
+        # and the file list to 20kB each.
+        STAT_LIMIT = 20 * 1024
+        DIFF_LIMIT = 20 * 1024
+
         # Get statistics for each file
         diff_stats = f"{self.pr.changed_files} Files Affected:\n\n"
         for file in self.pr.get_files():
@@ -125,35 +148,42 @@ def run(self) -> bool:
             if file.status == "renamed":
                 print(f"(from {file.previous_filename})")
             diff_stats += "\n"
-        diff_stats += "\n"
+            if len(diff_stats) > STAT_LIMIT:
+                break
 
         # Get the diff
         try:
-            patch = html.escape(requests.get(self.pr.diff_url).text)
+            patch = requests.get(self.pr.diff_url).text
         except:
             patch = ""
-        diff_stats += "\n<pre>\n" + html.escape(patch)
 
-        # GitHub limits comments to 65,536 characters, let's limit the diff to 20kB.
-        DIFF_LIMIT = 20 * 1024
+        patch = sanitize_markdown_code_block(patch)
+
         patch_link = f"Full diff: {self.pr.diff_url}\n"
         if len(patch) > DIFF_LIMIT:
             patch_link = f"\nPatch is {human_readable_size(len(patch))}, truncated to {human_readable_size(DIFF_LIMIT)} below, full version: {self.pr.diff_url}\n"
-            diff_stats = diff_stats[0:DIFF_LIMIT] + "...\n<truncated>\n"
-        diff_stats += "</pre>"
+            patch = patch[0:DIFF_LIMIT] + "...\n[truncated]\n"
         team_mention = "@llvm/{}".format(team.slug)
 
-        body = self.pr.body
+        body = escape_description(self.pr.body)
+        # Note: the comment is in markdown and the code below
+        # is sensible to line break
         comment = f"""
 {self.COMMENT_TAG}
 {team_mention}
-            
+
 <details>
 <summary>Changes</summary>
 {body}
 --
 {patch_link}
+
 {diff_stats}
+
+```diff
+{patch}
+```
+
 </details>
 """
 

```

</details>

cor3ntin committed:
add a line break to make sure the description is markdown-formatted

cor3ntin commented:
<!--LLVM PR SUMMARY COMMENT-->

@llvm/pr-subscribers-github-workflow

<details>
<summary>Changes</summary>

* This avoid pinging folks on all issue when they got pinged on bugzilla eons ago
* Avoid formatting bugs when there is html in the issue description
* Truncate the list of files and the diff independently of each other. This avoids truncating cutting a file line in 2 and to cut in the middle of html markup. This is a fringe case but it does happen when people accidentally push weird branches conflicting on all the files.


---
Full diff: https://github.com/llvm/llvm-project/pull/66118.diff


1 Files Affected:

- (modified) llvm/utils/git/github-automation.py (+47-16) 


```diff
diff --git a/llvm/utils/git/github-automation.py b/llvm/utils/git/github-automation.py
index eac5816b5499f6a..8578a83262717ab 100755
--- a/llvm/utils/git/github-automation.py
+++ b/llvm/utils/git/github-automation.py
@@ -47,6 +47,21 @@ def _get_curent_team(team_name, teams) -> Optional[github.Team.Team]:
     return None
 
 
+def escape_description(str):
+    # https://github.com/github/markup/issues/1168#issuecomment-494946168
+    str = html.escape(str, False)
+    # '@' followed by alphanum is a user name
+    str = re.sub("@(?=\w+)", "@<!-- -->", str)
+    # '#' followed by digits is considered an issue number
+    str = re.sub("#(?=\d+\s)", "#<!-- -->", str)
+    return str
+
+
+def sanitize_markdown_code_block(str):
+    # remove codeblocks terminators
+    return re.sub("^\s*```\s*$", r"` ` `", str)
+
+
 class IssueSubscriber:
     @property
     def team_name(self) -> str:
@@ -67,12 +82,15 @@ def run(self) -> bool:
         if team.slug == "issue-subscribers-good-first-issue":
             comment = "{}\n".format(beginner_comment)
 
-        comment = (
-            f"@llvm/{team.slug}"
-            + "\n\n<details>\n"
-            + f"{self.issue.body}\n"
-            + "</details>"
-        )
+        body = escape_description(self.issue.body)
+
+        comment = f"""
+@llvm/{team.slug}
+
+<details>
+{body}
+</details>
+"""
 
         self.issue.create_comment(comment)
         return True
@@ -113,6 +131,11 @@ def run(self) -> bool:
             print(f"couldn't find team named {self.team_name}")
             return False
 
+        # GitHub limits comments to 65,536 characters, let's limit the diff
+        # and the file list to 20kB each.
+        STAT_LIMIT = 20 * 1024
+        DIFF_LIMIT = 20 * 1024
+
         # Get statistics for each file
         diff_stats = f"{self.pr.changed_files} Files Affected:\n\n"
         for file in self.pr.get_files():
@@ -125,35 +148,43 @@ def run(self) -> bool:
             if file.status == "renamed":
                 print(f"(from {file.previous_filename})")
             diff_stats += "\n"
-        diff_stats += "\n"
+            if len(diff_stats) > STAT_LIMIT:
+                break
 
         # Get the diff
         try:
-            patch = html.escape(requests.get(self.pr.diff_url).text)
+            patch = requests.get(self.pr.diff_url).text
         except:
             patch = ""
-        diff_stats += "\n<pre>\n" + html.escape(patch)
 
-        # GitHub limits comments to 65,536 characters, let's limit the diff to 20kB.
-        DIFF_LIMIT = 20 * 1024
+        patch = sanitize_markdown_code_block(patch)
+
         patch_link = f"Full diff: {self.pr.diff_url}\n"
         if len(patch) > DIFF_LIMIT:
             patch_link = f"\nPatch is {human_readable_size(len(patch))}, truncated to {human_readable_size(DIFF_LIMIT)} below, full version: {self.pr.diff_url}\n"
-            diff_stats = diff_stats[0:DIFF_LIMIT] + "...\n<truncated>\n"
-        diff_stats += "</pre>"
+            patch = patch[0:DIFF_LIMIT] + "...\n[truncated]\n"
         team_mention = "@llvm/{}".format(team.slug)
 
-        body = self.pr.body
+        body = escape_description(self.pr.body)
+        # Note: the comment is in markdown and the code below
+        # is sensible to line break
         comment = f"""
 {self.COMMENT_TAG}
 {team_mention}
-            
+
 <details>
 <summary>Changes</summary>
+
 {body}
---
+---
 {patch_link}
+
 {diff_stats}
+
+```diff
+{patch}
+```
+
 </details>
 """
 

```

</details>

Endilll commented on llvm/utils/git/github-automation.py:
````diff
@@ -47,6 +47,21 @@ def _get_curent_team(team_name, teams) -> Optional[github.Team.Team]:
+
+def sanitize_markdown_code_block(str):
+    # remove codeblocks terminators
+    return re.sub("^\s*```\s*$", r"` ` `", str)
````
There could be code blocks that specify syntax in order to enable syntax highlighting, e.g. 
<pre>
```cpp
std::vector<int> v;
```
</pre>

cor3ntin replied:
we only need to escape the closing ` ``` `

```
```cpp does not close a code block 
```

Endilll replied:
Ah, I though you match both opening and closing triple backticks.
But is this going to work correctly for multiple code blocks in a message? Multiple unescaped openings might be interpreted as opening and closing.

cor3ntin replied:
Nope - but re-reading the spect and doing more tests, we can use more backticks instead of escaping https://spec.commonmark.org/0.30/#fenced-code-block

cor3ntin commented:
<!--LLVM PR SUMMARY COMMENT-->

@llvm/pr-subscribers-github-workflow

<details>
<summary>Changes</summary>

* This avoid pinging folks on all issue when they got pinged on bugzilla eons ago
* Avoid formatting bugs when there is html in the issue description
* Truncate the list of files and the diff independently of each other. This avoids truncating cutting a file line in 2 and to cut in the middle of html markup. This is a fringe case but it does happen when people accidentally push weird branches conflicting on all the files.

Test: This is #<!-- -->66118 @<!-- -->cor3ntin 

---
Full diff: https://github.com/llvm/llvm-project/pull/66118.diff


1 Files Affected:

- (modified) llvm/utils/git/github-automation.py (+47-16) 


```diff
diff --git a/llvm/utils/git/github-automation.py b/llvm/utils/git/github-automation.py
index eac5816b5499f6a..8578a83262717ab 100755
--- a/llvm/utils/git/github-automation.py
+++ b/llvm/utils/git/github-automation.py
@@ -47,6 +47,21 @@ def _get_curent_team(team_name, teams) -> Optional[github.Team.Team]:
     return None
 
 
+def escape_description(str):
+    # https://github.com/github/markup/issues/1168#issuecomment-494946168
+    str = html.escape(str, False)
+    # '@' followed by alphanum is a user name
+    str = re.sub("@(?=\w+)", "@<!-- -->", str)
+    # '#' followed by digits is considered an issue number
+    str = re.sub("#(?=\d+\s)", "#<!-- -->", str)
+    return str
+
+
+def sanitize_markdown_code_block(str):
+    # remove codeblocks terminators
+    return re.sub("^\s*```\s*$", r"` ` `", str)
+
+
 class IssueSubscriber:
     @property
     def team_name(self) -> str:
@@ -67,12 +82,15 @@ def run(self) -> bool:
         if team.slug == "issue-subscribers-good-first-issue":
             comment = "{}\n".format(beginner_comment)
 
-        comment = (
-            f"@llvm/{team.slug}"
-            + "\n\n<details>\n"
-            + f"{self.issue.body}\n"
-            + "</details>"
-        )
+        body = escape_description(self.issue.body)
+
+        comment = f"""
+@llvm/{team.slug}
+
+<details>
+{body}
+</details>
+"""
 
         self.issue.create_comment(comment)
         return True
@@ -113,6 +131,11 @@ def run(self) -> bool:
             print(f"couldn't find team named {self.team_name}")
             return False
 
+        # GitHub limits comments to 65,536 characters, let's limit the diff
+        # and the file list to 20kB each.
+        STAT_LIMIT = 20 * 1024
+        DIFF_LIMIT = 20 * 1024
+
         # Get statistics for each file
         diff_stats = f"{self.pr.changed_files} Files Affected:\n\n"
         for file in self.pr.get_files():
@@ -125,35 +148,43 @@ def run(self) -> bool:
             if file.status == "renamed":
                 print(f"(from {file.previous_filename})")
             diff_stats += "\n"
-        diff_stats += "\n"
+            if len(diff_stats) > STAT_LIMIT:
+                break
 
         # Get the diff
         try:
-            patch = html.escape(requests.get(self.pr.diff_url).text)
+            patch = requests.get(self.pr.diff_url).text
         except:
             patch = ""
-        diff_stats += "\n<pre>\n" + html.escape(patch)
 
-        # GitHub limits comments to 65,536 characters, let's limit the diff to 20kB.
-        DIFF_LIMIT = 20 * 1024
+        patch = sanitize_markdown_code_block(patch)
+
         patch_link = f"Full diff: {self.pr.diff_url}\n"
         if len(patch) > DIFF_LIMIT:
             patch_link = f"\nPatch is {human_readable_size(len(patch))}, truncated to {human_readable_size(DIFF_LIMIT)} below, full version: {self.pr.diff_url}\n"
-            diff_stats = diff_stats[0:DIFF_LIMIT] + "...\n<truncated>\n"
-        diff_stats += "</pre>"
+            patch = patch[0:DIFF_LIMIT] + "...\n[truncated]\n"
         team_mention = "@llvm/{}".format(team.slug)
 
-        body = self.pr.body
+        body = escape_description(self.pr.body)
+        # Note: the comment is in markdown and the code below
+        # is sensible to line break
         comment = f"""
 {self.COMMENT_TAG}
 {team_mention}
-            
+
 <details>
 <summary>Changes</summary>
+
 {body}
---
+---
 {patch_link}
+
 {diff_stats}
+
+```diff
+{patch}
+```
+
 </details>
 """
 

```

</details>

cor3ntin committed:
Use more backticks to start the code block

That way we do not have to escape ` ``` `

jayfoad commented on llvm/utils/git/github-automation.py:
```diff
@@ -125,35 +143,41 @@ def run(self) -> bool:
+
 {body}
---
+---
```
This change (three dashes) causes the last paragraph of the body to be rendered like a title, which looks awful, e.g.: the "After legalization is done" paragraph in this issue: https://github.com/llvm/llvm-project/pull/66567#issuecomment-1722108154

cor3ntin replied:
We need to add an extra new line before the dashes, I'll do that

cor3ntin replied:
Fixed
